libraries/WebServer/src/Parsing.cpp

```diff
@@ -304,11 +304,42 @@ void WebServer::_uploadWriteByte(uint8_t b){
 
 int WebServer::_uploadReadByte(WiFiClient& client){
   int res = client.read();
-  if(res == -1){
-    while(!client.available() && client.connected())
-      delay(2);
-    res = client.read();
+  if(res < 0) {
+    // keep trying until you either read a valid byte or timeout
+    unsigned long startMillis = millis();
+    long timeoutIntervalMillis = client.getTimeout();
+    unsigned long currentMillis = startMillis;
+    boolean timedOut = false;
+    for(;;) {
+      // loosely modeled after blinkWithoutDelay pattern
+      while(!timedOut && !client.available() && client.connected()){
+        delay(2);
+        currentMillis = millis();
+        timedOut = currentMillis - startMillis >= timeoutIntervalMillis;
```

Comment:
I would say, do not use at all `currentMillis` and just replace it with a call to millis(). The compiler will do it underneath anyway, but I think it will make the code slightly more readable

Reply:
I'd be happy to _not_ use a local variable, but I think we have to at least cast the result to (unsigned long) for the comparison to work and still handle millis() overflow boundary conditions

Reply:
addressed in d5c501f